src/java.base/windows/classes/sun/nio/fs/WindowsFileSystemProvider.java

```diff
@@ -374,11 +374,25 @@ public void checkAccess(Path obj, AccessMode... modes) throws IOException {
             }
         }
 
-        // special-case read access to avoid needing to determine effective
-        // access to file; default if modes not specified
         if (!w && !x) {
-            checkReadAccess(file);
-            return;
+            try {
+                // special-case read access to avoid needing to determine
+                // effective access to file
+                checkReadAccess(file);
+                return;
+            } catch (IOException ioe) {
+                if (!r) {
+                    // Modes not specified: existence check only.
+                    // Assume file exists if attributes can be read.
+                    try {
+                        WindowsFileAttributes.get(file, true);
+                        return;
+                    } catch (WindowsException e) {
+                        ioe.addSuppressed(e);
+                    }
+                }
+                throw ioe;
+            }
```

Comment:
Testing the existence of a file is performance issue in some usages and would be interesting to see if this would be a bit better.

```
        // check file exists only
        if (!r && !w && !x) {
            file.checkRead();
            try {
                WindowsFileAttributes.get(file, true);
                return;
            } catch (WindowsException e) {
                exc.rethrowAsIOException(file);
            }
        }
```

Reply:
That was actually my first draft version. What happened to `checkReadAccess()` here?

Reply:
checkReadAccess comes later, when !w & !x.

Reply:
I misread that. I'll give it a try.